sql/src/main/java/io/crate/execution/jobs/TasksService.java

```diff
@@ -58,7 +58,7 @@
     private final ConcurrentMap<UUID, RootTask> activeTasks =
         ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
 
-    private final List<KillAllListener> killAllListeners = Collections.synchronizedList(new ArrayList<>());
+    private final List<KillAllListener> killAllListeners = new CopyOnWriteArrayList<>();
```

Comment:
hm, shall we do this? with synchronizedList we're in danger of ConcurrentModificationException while we iterated over the listeners so we do need to fix this, but is COWAL the answer? (we mostly write in this list and COWAL is way slower and creates garbage) IMO we should keep synchronizedList and create a manual copy only when we do iterate over it and kill all the listeners in this "copy list"

Reply:
thinking about it some more, the size of this list is usually relatively small so maybe the COWAL overhead is not significant

Reply:
See the description in the commit message and PR description. The writes only happen once on node start-up (and I think currently it is just 1 write)

Reply:
yep, makes sense (didn't see the @Singleton/Insert on TransportShardUpsertAction), good catch (lock free reads ❤️ )